### DISCUSSION TITLE
How do you read text from a project file?

### DISCUSSION THREAD

**Discussion by MarkBrainfire** [AvaloniaUI/Avalonia#12710] [General]
I want to include a text file in my project and then at runtime read the text in this file into a string variable.

How do you do this?

Mark

**Comment by wieslawsoltes** [comment]
https://github.com/AvaloniaUI/XamlPlayground/blob/6f5ceb9dbfb19c3024c84c517af2c2e2b2f3ef54/src/XamlPlayground/XamlPlayground.csproj#L30

https://github.com/AvaloniaUI/XamlPlayground/blob/6f5ceb9dbfb19c3024c84c517af2c2e2b2f3ef54/src/XamlPlayground/ViewModels/MainViewModel.cs#L148-L173

**Reply by MarkBrainfire** [reply]
Thank you for your reply @wieslawsoltes.  I've looked at the Playground links but still don't understand.  

Reading text from a file or URI or from a text file embedded in the project ought to be a simple thing to do, but for some reason it is difficult.  I'll keep studying and trying to figure all of this out.

**Comment by timunie** [comment]
checkout AssetLoader static methods

**Reply by timunie** [reply]
https://docs.avaloniaui.net/docs/next/basics/user-interface/assets

**Comment by MarkBrainfire** [comment]
Thanks Timunie. The documentation only talks about image file assets. I'll try to figure out how to deal with text file assets and then how to read the text from these files.

**Reply by stevemonaco** [reply]
It's not much different. You get a `Stream` as a result of `AssetLoader.Open` and then you read it as if you're reading text from any `Stream`. eg.

```cs
public partial class MainWindowViewModel : ViewModelBase
{
    public string Greeting { get; }

    public MainWindowViewModel()
    {
        using var resource = AssetLoader.Open(new Uri(@"avares://AvaloniaApplication24/Assets/Testing.txt"));
        using var reader = new StreamReader(resource);

        var contents = reader.ReadToEnd();
        Greeting = contents;
    }
}
```

This is for demonstration purposes. You should be doing this logic in code-behind instead of the VM (as `AssetLoader` is coupled to Avalonia). Be sure that your text file has the Build Action of `AvaloniaResource`. The Avalonia project templates include everything in the Assets folder as one by default.

**Reply by MarkBrainfire** [reply]
Thanks Stephen,
I'm grateful for your response.  I just worked out essentially the same solution you provided.  

Copy your text file to the Assets folder.  In VS set its Build Action to AvaloniaResource and Copy to Output Drive to Copy if newer.  Then it's easy to retrieve the text:

    private static string GetAssetsTextfileText(string fileName)
    {
        Uri textFileUri = new ("avares://MyProject/Assets/" + fileName);
        using StreamReader streamReader = new (AssetLoader.Open(textFileUri));
        return streamReader.ReadToEnd();
    }
